### PR TITLE
JAMES-1702 Remove Bcc field when sending a mail

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

**Comment by chibenwa** [comment]
I would love to see tests validating that for SMTP as well.

**Comment by rouazana** [comment]
:+1:

**Comment by mbaechler** [comment]
You should all be afraid how much this is not DRY. And you brittle it is.
Mailet configuration is a sample, you have no guarantee about what will be deployed.
If you want to use mailets, you need to ensure the mailet is always installed when jmap is installed.

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java [hunk 1]
```diff
@@ -1059,6 +1059,190 @@ public void setMessagesShouldDeliverMessageToRecipient() throws Exception {
...
+            .contentType(ContentType.JSON)
+            .header("Authorization", recipientToken.serialize())
+            .body("[[\"getMessageList\", {\"fetchMessages\": true, \"fetchMessageProperties\": [\"textBody\", \"bcc\"] }, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .log().ifValidationFails()
+            .statusCode(200)
+            .body(SECOND_NAME, equalTo("messages"))
+            .body(SECOND_ARGUMENTS + ".list[0].bcc", empty());
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: why `.list[0]` ?

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java [hunk 2]
```diff
@@ -1059,6 +1059,190 @@ public void setMessagesShouldDeliverMessageToRecipient() throws Exception {
...
+        .when()
+                .post("/jmap");
+
+        // Then
+        calmlyAwait.atMost(10, TimeUnit.SECONDS).until( () -> isAnyMessageFoundInRecipientsMailboxes(recipientToken));
+        with()
+            .accept(ContentType.JSON)
+            .contentType(ContentType.JSON)
+            .header("Authorization", recipientToken.serialize())
+            .body("[[\"getMessageList\", {\"fetchMessages\": true, \"fetchMessageProperties\": [\"textBody\", \"bcc\"] }, \"#0\"]]")
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: you probably don't need textBody

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java [hunk 3]
```diff
@@ -1059,6 +1059,190 @@ public void setMessagesShouldDeliverMessageToRecipient() throws Exception {
...
+        given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .header("Authorization", this.accessToken.serialize())
+                .body(requestBody)
+        // When
+        .when()
+                .post("/jmap");
+
+        // Then
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: or not

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java [hunk 4]
```diff
@@ -1059,6 +1059,190 @@ public void setMessagesShouldDeliverMessageToRecipient() throws Exception {
...
+            .post("/jmap")
+        .then()
+            .log().ifValidationFails()
+            .statusCode(200)
+            .body(SECOND_NAME, equalTo("messages"))
+            .body(SECOND_ARGUMENTS + ".list[0].bcc", hasSize(1));
+    }
+
+    @Test
+    public void setMessagesShouldSendMessageToBcc() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: how is this test different from stripBcc one ?

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: In order to retrieve the first message in the list.

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: There are thrre different tests:
- check that a mail with a _To_ and a _Bcc_ is sent to _To_ without bcc field
- check that a mail with a _To_ and a _Bcc_ is stored in _From_ sent mailbox with bcc field
- check that a mail with a _To_ and a _Bcc_ is sent to _Bcc_ without bcc

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 1 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: Ok but you don't assert there's a single message first

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 4 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: ok, understood

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: OK

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/SetMessagesMethodTest.java]: ?

**Comment by aduprat** [comment]
- check matcher
- better error message

**Comment by rouazana** [comment]
:+1:

**Pull request merged**